cmd/gardener-node-agent/app/app.go

```diff
@@ -0,0 +1,227 @@
+func NewCommand() *cobra.Command {
+	opts := &options{}
+
+	cmd := &cobra.Command{
+		Use:   Name,
+		Short: "Launch the " + Name,
+		Args:  cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			verflag.PrintAndExitIfRequested()
+
+			if err := opts.complete(); err != nil {
+				return err
+			}
+			if err := opts.validate(); err != nil {
+				return err
+			}
+
+			log, err := logger.NewZapLogger(opts.config.LogLevel, opts.config.LogFormat)
+			if err != nil {
+				return fmt.Errorf("error instantiating zap logger: %w", err)
+			}
+
+			logf.SetLogger(log)
+			klog.SetLogger(log)
+
+			log.Info("Starting "+Name, "version", version.Get())
+			cmd.Flags().VisitAll(func(flag *pflag.Flag) {
+				log.Info(fmt.Sprintf("FLAG: --%s=%s", flag.Name, flag.Value)) //nolint:logcheck
+			})
+
+			// don't output usage on further errors raised during execution
+			cmd.SilenceUsage = true
+			// further errors will be logged properly, don't duplicate
+			cmd.SilenceErrors = true
+
+			return run(cmd.Context(), log, opts.config)
+		},
+	}
+
+	flags := cmd.Flags()
+	verflag.AddFlags(flags)
+	opts.addFlags(flags)
+
+	return cmd
+}
+
+func run(ctx context.Context, log logr.Logger, cfg *config.NodeAgentConfiguration) error {
+	log.Info("Feature Gates", "featureGates", features.DefaultFeatureGate)
+
+	log.Info("Getting rest config")
+	if kubeconfig := os.Getenv("KUBECONFIG"); kubeconfig != "" {
+		cfg.ClientConnection.Kubeconfig = kubeconfig
+	}
+	if cfg.ClientConnection.Kubeconfig == "" {
+		return fmt.Errorf("must specify path to a kubeconfig (either via \"KUBECONFIG\" environment variable of via .clientConnection.kubeconfig in component config)")
+	}
+
+	restConfig, err := kubernetes.RESTConfigFromClientConnectionConfiguration(&cfg.ClientConnection, nil, kubernetes.AuthTokenFile)
+	if err != nil {
+		return err
+	}
+
+	if restConfig.BearerTokenFile == nodeagentv1alpha1.BootstrapTokenFilePath {
+		log.Info("Kubeconfig points to the bootstrap token file")
+		if err := fetchAccessTokenViaBootstrapToken(ctx, log, restConfig, cfg); err != nil {
+			return fmt.Errorf("failed fetching access token via bootstrap token: %w", err)
+		}
+	}
+
+	var extraHandlers map[string]http.Handler
+	if cfg.Debugging != nil && cfg.Debugging.EnableProfiling {
+		extraHandlers = routes.ProfilingHandlers
+		if cfg.Debugging.EnableContentionProfiling {
+			goruntime.SetBlockProfileRate(1)
+		}
+	}
+
+	log.Info("Setting up manager")
+	mgr, err := manager.New(restConfig, manager.Options{
+		Logger:                  log,
+		Scheme:                  kubernetes.SeedScheme,
+		GracefulShutdownTimeout: pointer.Duration(5 * time.Second),
+
+		HealthProbeBindAddress: net.JoinHostPort(cfg.Server.HealthProbes.BindAddress, strconv.Itoa(cfg.Server.HealthProbes.Port)),
+		Metrics: metricsserver.Options{
+			BindAddress:   net.JoinHostPort(cfg.Server.Metrics.BindAddress, strconv.Itoa(cfg.Server.Metrics.Port)),
+			ExtraHandlers: extraHandlers,
+		},
```

Comment:
There is neither a validation nor a defaulting for health probe and metrics port. Thus, there is a risk of creating node agent configs which make it fail on startup.

And just for confirmation. Node agent is opening a node port here. It is safe to assume that the node network is not publicly available, isn't it?

Reply:
> There is neither a validation nor a defaulting for health probe and metrics port. Thus, there is a risk of creating node agent configs which make it fail on startup.

`gardenlet` will be the only user of GNA and its configuration - we probably won't deploy it with "broken" configuration. If end-users manually tamper with it - sure, they can break it, just like they can break anything else as well.

> And just for confirmation. Node agent is opening a node port here. It is safe to assume that the node network is not publicly available, isn't it?

Even if not, having the `kubelet` publicly available is probably the greater risk compared to the GNA health/metrics endpoint 😄